compiler/rustc_const_eval/src/transform/validate.rs

```diff
@@ -351,10 +341,18 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                     self.fail(location, format!("bad arg ({:?} != usize)", op_cnt_ty))
                 }
             }
-            StatementKind::SetDiscriminant { .. }
-            | StatementKind::StorageLive(..)
+            StatementKind::SetDiscriminant { .. } => {
+                if self.mir_phase < MirPhase::DropLowering {
+                    self.fail(location, "`SetDiscriminant` is not allowed until drop elaboration");
+                }
+            }
+            StatementKind::Retag(_, _) => {
+                // FIXME(JakobDegen) The validator should check that `self.mir_phase <
+                // DropLowering`. However, this causes ICEs with generation of drop shims, which
+                // seem to fail to set their `MirPhase` correctly.
```

Comment:
If anyone familiar with the relevant code can lend some guidance here, I'd be happy to add a fix to this PR. I don't understand this code well enough to be comfortable adding it without further consultation though. For example, one of the things I am confused about is why this doesn't cause more ICEs

Reply:
retag statements are only emitted when you add `-Zmir-emit-retag`, which only has two mir-opt tests.

Reply:
This happens after drop elaboration, so at best you can forbid it after Deaggregator

Reply:
You mean before? The problem isn't the `Retag` statements in "normal" code, the problem is that drop shim generation creates and verifies a body with `MirPhase::Building` that contains `Retag` statements (when `-Zmir-emit-retag` is set)

Reply:
oh wow, yea I did not think about shims at all. I guess we'll just allow retags everywhere?

Reply:
We could, but I'd like to leave this as a fixme because I don't think we should decide which dialect of MIR drop shims use based on this detail. I don't understand the code well enough to do this right now, but in the future I would like to go through this and ensure that shims cleanly adhere to the semantics of whatever dialect of MIR they claim to be using.